OsmAnd/src/net/osmand/plus/routing/SuggestionsMapsProvider.java

```diff
@@ -0,0 +1,145 @@
+				Location location = calculateMidPoint(points.get(i), points.get(nextIndex));
+				points.add(nextIndex, location);
+			}
+		}
+		return points;
+	}
+
+	public LinkedList<Location> getStartFinishIntermediatesPoints(String locationProvider) {
+		LinkedList<Location> points = new LinkedList<>();
+		points.add(new Location(locationProvider, params.start.getLatitude(), params.start.getLongitude()));
+		if (params.intermediates != null) {
+			for (LatLon l : params.intermediates) {
+				points.add(new Location(locationProvider, l.getLatitude(), l.getLongitude()));
+			}
+		}
+		points.add(new Location(locationProvider, params.end.getLatitude(), params.end.getLongitude()));
+		return points;
+	}
+
+	public List<Location> findOnlineRoutePoints() throws IOException, JSONException {
+		List<LatLon> points = new ArrayList<>();
+		List<Location> route;
+		List<Location> routeLocation = new ArrayList<>();
+		OsmandApplication app = params.ctx;
+		OnlineRoutingHelper helper = app.getOnlineRoutingHelper();
+		LinkedList<Location> location = getStartFinishIntermediatesPoints("");
+		for (Location e : location) {
+			points.add(new LatLon(e.getLatitude(), e.getLongitude()));
+		}
+		OnlineRoutingEngine.OnlineRoutingResponse response =
+				helper.calculateRouteOnline(createInitStateEngine(), points, params.leftSide);
+		if (response != null) {
+			route = response.getRoute();
+			routeLocation.addAll(route);
+		}
+		return getOnlineLocationsBasedOnDistance(routeLocation);
+	}
+
+	public List<WorldRegion> getMissingMaps(List<Location> points) throws IOException {
+		LinkedHashSet<WorldRegion> suggestedMaps = new LinkedHashSet<>();
+		for (int i = 0; i < points.size(); i++) {
+			final Location o = points.get(i);
+			LatLon latLonPoint = new LatLon(o.getLatitude(), o.getLongitude());
+			List<WorldRegion> downloadRegions = params.ctx.getRegions().getWorldRegionsAt(latLonPoint);
+
+			if (downloadRegions.isEmpty()) {
+				pointOnWater = true;
+			}
+
+			boolean addMaps = true;
+			List<WorldRegion> maps = new ArrayList<>();
+			for (WorldRegion downloadRegion : downloadRegions) {
+				String mapName = downloadRegion.getRegionDownloadName();
+				String countryMapName = downloadRegion.getSuperregion().getRegionDownloadName();
+				List<WorldRegion> subregions = downloadRegion.getSubregions();
+				boolean isSuggestedMapsNeeded = !checkIfObjectDownloaded(mapName);
+				boolean isCountry = Algorithms.isEmpty(countryMapName) && !subregions.isEmpty();
+				if (isSuggestedMapsNeeded) {
+					if (!isCountry) {
+						maps.add(downloadRegion);
+					}
+				} else {
+					addMaps = false;
+				}
+			}
+			if (addMaps) {
+				suggestedMaps.addAll(maps);
+			}
+		}
+		return new ArrayList<>(suggestedMaps);
+	}
+
+	@NonNull
+	private List<Location> getOnlineLocationsBasedOnDistance(List<Location> routeLocation) {
+		List<Location> mapsBasedOnPoints = new ArrayList<>();
+		for (int i = 0, j; i < routeLocation.size(); i = j) {
```

Comment:
this an impossible loop with all variables messed up, it is not clear what's going on here